File: tools-manage/invocationlab-admin/src/main/java/org/javamaster/invocationlab/admin/util/SerializationUtils.java
package org.javamaster.invocationlab.admin.util;

import org.javamaster.invocationlab.admin.config.BizException;
import org.javamaster.invocationlab.admin.service.Pair;
import org.javamaster.invocationlab.admin.service.creation.impl.JdkCreator;
import lombok.Cleanup;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.apache.tomcat.util.buf.HexUtils;
import org.springframework.core.ConfigurableObjectInputStream;
import org.springframework.core.convert.ConversionService;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.nio.charset.StandardCharsets;

@Slf4j
public class SerializationUtils {
    private static final byte AC = HexUtils.fromHexString("AC")[0];
    private static final byte ED = HexUtils.fromHexString("ED")[0];

    @SneakyThrows
    public static byte[] serialize(Object object) {
        if (!(object instanceof Serializable)) {
            throw new IllegalArgumentException(object.getClass().getSimpleName() + " requires a Serializable payload " +
                    "but received an object of type [" + object.getClass().getName() + "]");
        }
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        @Cleanup
        ObjectOutputStream objectOutputStream = new ObjectOutputStream(outputStream);
        objectOutputStream.writeObject(object);
        objectOutputStream.flush();
        return outputStream.toByteArray();
    }

    @SneakyThrows
    public static Object deserialize(byte[] bytes, ClassLoader classLoader) {
        @Cleanup
        InputStream inputStream = new ByteArrayInputStream(bytes);
        ObjectInputStream objectInputStream = new ConfigurableObjectInputStream(inputStream, classLoader);
        return objectInputStream.readObject();
    }

    public static Pair<String, Class<?>> dealJdkDeserialize(byte[] bytes) {
        String val;
        Class<?> clazz = Void.class;
        try {
            Object deserializeObj = SerializationUtils.deserialize(bytes, JdkCreator.globalJdkClassLoader);
            clazz = deserializeObj.getClass();
            if (ClassUtils.isPrimitive(clazz.getTypeName())) {
                val = deserializeObj.toString();
            } else {
                val = JsonUtils.jacksonObjectMapper.writeValueAsString(deserializeObj);
            }
        } catch (Exception e) {
            String value = new String(bytes, StandardCharsets.UTF_8);
            log.error(value, e);
            val = "无法反序列化" + e.getMessage() + ",缺少class,请先在解析Redis JDK序列化依赖菜单解析依赖!!!\r\n\r\n" + value;
        }
        return new Pair<>(val, clazz);
    }

    @SneakyThrows
    public static Object convertValToObj(String val, Class<?> clazz) {
        if (clazz == Void.class) {
            clazz = JsonUtils.isJsonObjOrArray(val) ? Object.class : String.class;
        }

        if (ClassUtils.isPrimitive(clazz.getTypeName())) {
            ConversionService conversionService = SpringUtils.getContext().getBean(ConversionService.class);
            return conversionService.convert(val, clazz);
        } else {
            ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
            try {
                Thread.currentThread().setContextClassLoader(JdkCreator.globalJdkClassLoader);
                Object obj = JsonUtils.jacksonObjectMapper.readValue(val, Object.class);
                if (obj == null) {
                    throw new BizException("无法反序列化!!!");
                }
                return obj;
            } catch (Exception e) {
                throw new RuntimeException(e);
            } finally {
                Thread.currentThread().setContextClassLoader(contextClassLoader);
            }
        }
    }

    public static boolean isJdkSerialize(byte[] bytes) {
        return bytes.length > 1 && bytes[0] == AC && bytes[1] == ED;
    }
}

File: tools-manage/invocationlab-admin-21/src/main/kotlin/org/javamaster/invocationlab/admin/util/SerializationUtils.kt
package org.javamaster.invocationlab.admin.util

import org.javamaster.invocationlab.admin.config.BizException
import org.javamaster.invocationlab.admin.service.Pair
import org.javamaster.invocationlab.admin.service.creation.impl.JdkCreator
import org.javamaster.invocationlab.admin.util.ClassUtils.isPrimitive
import org.javamaster.invocationlab.admin.util.JsonUtils.isJsonObjOrArray
import org.apache.tomcat.util.buf.HexUtils
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.core.ConfigurableObjectInputStream
import org.springframework.core.convert.ConversionService
import java.io.*
import java.nio.charset.StandardCharsets


object SerializationUtils {
    private val log: Logger = LoggerFactory.getLogger(this.javaClass)
    private val AC = HexUtils.fromHexString("AC")[0]
    private val ED = HexUtils.fromHexString("ED")[0]

    @JvmStatic

    fun serialize(`object`: Any): ByteArray {
        require(`object` is Serializable) {
            `object`.javaClass.simpleName + " requires a Serializable payload " +
                    "but received an object of type [" + `object`.javaClass.name + "]"
        }
        val outputStream = ByteArrayOutputStream()
        ObjectOutputStream(outputStream).use {
            it.writeObject(`object`)
            it.flush()
            return outputStream.toByteArray()
        }
    }

    fun deserialize(bytes: ByteArray?, classLoader: ClassLoader?): Any {
        ByteArrayInputStream(bytes).use {
            val objectInputStream: ObjectInputStream = ConfigurableObjectInputStream(it, classLoader)
            return objectInputStream.readObject()
        }
    }

    @JvmStatic
    fun dealJdkDeserialize(bytes: ByteArray): Pair<String, Class<*>> {
        var `val`: String
        var clazz: Class<*> = Void::class.java
        try {
            val deserializeObj = deserialize(bytes, JdkCreator.globalJdkClassLoader)
            clazz = deserializeObj.javaClass
            `val` = if (isPrimitive(clazz.typeName)) {
                deserializeObj.toString()
            } else {
                JsonUtils.jacksonObjectMapper.writeValueAsString(deserializeObj)
            }
        } catch (e: Exception) {
            val value = String(bytes, StandardCharsets.UTF_8)
            log.error(value, e)
            `val` = """
                无法反序列化${e.message},缺少class,请先在解析Redis JDK序列化依赖菜单解析依赖!!!
                
                $value
                """.trimIndent()
        }
        return Pair(`val`, clazz)
    }

    @JvmStatic

    fun convertValToObj(`val`: String?, clazz: Class<*>): Any? {
        var cls = clazz
        if (cls == Void::class.java) {
            cls = if (isJsonObjOrArray(`val`!!)) Any::class.java else String::class.java
        }

        if (isPrimitive(cls.typeName)) {
            val conversionService = SpringUtils.context.getBean(
                ConversionService::class.java
            )
            return conversionService.convert(`val`, cls)
        } else {
            val contextClassLoader = Thread.currentThread().contextClassLoader
            try {
                Thread.currentThread().contextClassLoader = JdkCreator.globalJdkClassLoader
                val obj = JsonUtils.jacksonObjectMapper.readValue(`val`, Any::class.java)
                    ?: throw BizException("无法反序列化!!!")
                return obj
            } catch (e: Exception) {
                throw RuntimeException(e)
            } finally {
                Thread.currentThread().contextClassLoader = contextClassLoader
            }
        }
    }

    @JvmStatic
    fun isJdkSerialize(bytes: ByteArray): Boolean {
        return bytes.size > 1 && bytes[0] == AC && bytes[1] == ED
    }
}
